client/client-multi/client-c/src/kaa/kaa_notification_manager.c

```diff
@@ -0,0 +1,1001 @@
+                kaa_list_remove_at(&self->optional_listeners, optional_listener_node, &kaa_optional_notification_wrapper_destroy);
+            }
+            outdated_topics = kaa_list_next(outdated_topics);
+        }
+    }
+    if (self->topics) {
+        kaa_list_destroy(self->topics, &kaa_topic_destroy);
+    }
+    self->topics = topics;
+    return kaa_notify_topic_update_subscribers(self ,topics);
+}
+
+kaa_error_t kaa_notification_received(kaa_notification_manager_t *self, kaa_list_t *notifications)
+{
+    KAA_RETURN_IF_NIL2(self, notifications, KAA_ERR_BADDATA);
+    kaa_topic_t* topic = NULL;
+    while (notifications) {
+        kaa_notification_wrapper_t *wrapper = kaa_list_get_data(notifications);
+        kaa_error_t err = kaa_find_topic(self, &topic, &wrapper->topic_id);
+        if (err) {
+            KAA_LOG_WARN(self->logger, KAA_ERR_NOT_FOUND, "Unknown notification received(topic id=%d) ", wrapper->topic_id);
+        }
+            if (kaa_notify_optional_notification_subscribers(self, &wrapper->topic_id, wrapper->notification_body)) {
+                kaa_notify_mandatory_notification_subscribers(self, &wrapper->topic_id, wrapper->notification_body);
+            }
+            notifications = kaa_list_next(notifications);
+    }
+    return KAA_ERR_NONE;
+}
+kaa_error_t kaa_notification_manager_handle_server_sync(kaa_notification_manager_t *self
+                                                      , kaa_platform_message_reader_t *reader
+                                                      , uint32_t extension_length)
+{
+    KAA_RETURN_IF_NIL2(self, reader, KAA_ERR_BADPARAM);
+    kaa_error_t err = KAA_ERR_NONE;
+    if (extension_length > 0) {
+        self->status->notification_seq_n = KAA_NTOHL(*((uint32_t *) reader->current));
+        reader->current += sizeof(uint32_t);
+        extension_length -= sizeof(uint32_t);
+        uint8_t field_id = 0;
+        while(extension_length > 0) {
+            field_id = *((uint8_t *) reader->current);
+            reader->current += sizeof(uint16_t);
+            extension_length -= sizeof(uint16_t);
+            switch (field_id) {
+            case NOTIFICATIONS: {
+                KAA_LOG_TRACE(self->logger, KAA_ERR_NONE, "Notifications received");
+                kaa_list_t *notifications = NULL;
+                kaa_notification_wrapper_t *notification = NULL;
+                uint16_t uid_length = 0;
+                uint16_t notifications_count = KAA_NTOHS(*((uint16_t *) reader->current));
+                reader->current += sizeof(uint16_t);
+                extension_length -= sizeof(uint16_t);
+                KAA_LOG_TRACE(self->logger, KAA_ERR_NONE, "Notifications count is %u", notifications_count);
+                if (notifications_count) {
+                    while (notifications_count--) {
+                        notification = (kaa_notification_wrapper_t *) KAA_MALLOC(sizeof(kaa_notification_wrapper_t));
```

Comment:
There is no need to copy all notification data. All what we need to give to the user is the deserialized notification body.
In your implementation there is unnecessary memory overhead.